From 0ca2e56b13c6da5ff09862a8345e7cd0bbf92d58 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Bernhard=20=C3=9Cbelacker?= <bernhardu@mailbox.org>
Date: Tue, 10 Dec 2024 11:33:31 +0100
Subject: [PATCH 81/86] windowscodecs: Workaround by increasing buffer.

=================================================================
==716==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x7f0814161054 at pc 0x6ffff67f2579 bp 0x7ffffe1fd980 sp 0x7ffffe1fd9c8
READ of size 1 at 0x7f0814161054 thread T0
02b8:fixme:file:server_get_file_info Unsupported info class e
    #0 0x6ffff67f2578 in tiff_decoder_read_tile /home/bernhard/data/entwicklung/2024/wine\wine/dlls/windowscodecs/libtiff.c:796:27
    #1 0x6ffff67f2578 in tiff_decoder_copy_pixels /home/bernhard/data/entwicklung/2024/wine\wine/dlls/windowscodecs/libtiff.c:954:22
    #2 0x6ffff67b0e27 in CommonDecoderFrame_CopyPixels /home/bernhard/data/entwicklung/2024/wine\wine/dlls/windowscodecs/decoder.c:423:10
    #3 0x0001401023c3 in IWICBitmapFrameDecode_CopyPixels /home/bernhard/data/entwicklung/2024/wine/wine-build/build-asan-pe/64/obj\include\wincodec.h:2018:12
    #4 0x0001401023c3 in test_color_formats /home/bernhard/data/entwicklung/2024/wine\wine/dlls/windowscodecs/tests/tiffformat.c:1247:14
    #5 0x0001401023c3 in func_tiffformat /home/bernhard/data/entwicklung/2024/wine\wine/dlls/windowscodecs/tests/tiffformat.c:1339:5
    #6 0x000140113d31 in run_test /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h:765:5
    #7 0x000140113d31 in main /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h:884:12
    #8 0x000140115c7f in mainCRTStartup /home/bernhard/data/entwicklung/2024/wine\wine/dlls/msvcrt/crt_main.c:58:11
    #9 0x6ffffbfd4808 in BaseThreadInitThunk /home/bernhard/data/entwicklung/2024/wine/wine/dlls/kernel32\thread.c:61:5
    #10 0x6ffffa8afa1a in RtlUserThreadStart (C:\windows\system32\ntdll.dll+0x17000fa1a)

0x7f0814161054 is located 0 bytes after 4-byte region [0x7f0814161050,0x7f0814161054)
allocated by thread T0 here:
    #0 0x6ffff84ba2c1 in malloc /home/bernhard/data/entwicklung/2024/llvm-mingw/2024-10-18/llvm-mingw/llvm-project/compiler-rt\lib/asan/asan_malloc_win.cpp:80:3
    #1 0x6ffff67f096d in tiff_decoder_read_tile /home/bernhard/data/entwicklung/2024/wine\wine/dlls/windowscodecs/libtiff.c:783:19
    #2 0x6ffff67f096d in tiff_decoder_copy_pixels /home/bernhard/data/entwicklung/2024/wine\wine/dlls/windowscodecs/libtiff.c:954:22
    #3 0x6ffff67b0e27 in CommonDecoderFrame_CopyPixels /home/bernhard/data/entwicklung/2024/wine\wine/dlls/windowscodecs/decoder.c:423:10
    #4 0x0001401023c3 in IWICBitmapFrameDecode_CopyPixels /home/bernhard/data/entwicklung/2024/wine/wine-build/build-asan-pe/64/obj\include\wincodec.h:2018:12
    #5 0x0001401023c3 in test_color_formats /home/bernhard/data/entwicklung/2024/wine\wine/dlls/windowscodecs/tests/tiffformat.c:1247:14
    #6 0x0001401023c3 in func_tiffformat /home/bernhard/data/entwicklung/2024/wine\wine/dlls/windowscodecs/tests/tiffformat.c:1339:5
    #7 0x000140113d31 in run_test /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h:765:5
    #8 0x000140113d31 in main /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h:884:12
    #9 0x000140115c7f in mainCRTStartup /home/bernhard/data/entwicklung/2024/wine\wine/dlls/msvcrt/crt_main.c:58:11
    #10 0x6ffffbfd4808 in BaseThreadInitThunk /home/bernhard/data/entwicklung/2024/wine/wine/dlls/kernel32\thread.c:61:5
    #11 0x6ffffa8afa1a in RtlUserThreadStart (C:\windows\system32\ntdll.dll+0x17000fa1a)

SUMMARY: AddressSanitizer: heap-buffer-overflow /home/bernhard/data/entwicklung/2024/wine\wine/dlls/windowscodecs/libtiff.c:796:27 in tiff_decoder_read_tile
Shadow bytes around the buggy address:
  0x7f0814160d80: fa fa fd fa fa fa fd fd fa fa fd fa fa fa fd fa
  0x7f0814160e00: fa fa fd fa fa fa fd fa fa fa fd fa fa fa fd fa
  0x7f0814160e80: fa fa fd fa fa fa 00 00 fa fa 00 00 fa fa 00 00
  0x7f0814160f00: fa fa fd fa fa fa fd fd fa fa 04 fa fa fa 04 fa
  0x7f0814160f80: fa fa 04 fa fa fa fd fa fa fa 00 fa fa fa fd fa
=>0x7f0814161000: fa fa 02 fa fa fa 00 fa fa fa[04]fa fa fa fa fa
  0x7f0814161080: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x7f0814161100: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x7f0814161180: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x7f0814161200: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x7f0814161280: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==716==ABORTING
02b8:fixme:kernelbase:AppPolicyGetProcessTerminationMethod FFFFFFFFFFFFFFFA, 00007FFFFEA8FE80
make: *** [Makefile:423367: dlls/windowscodecs/tests/x86_64-windows/tiffformat.ok] Fehler 1
---
 dlls/windowscodecs/libtiff.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dlls/windowscodecs/libtiff.c b/dlls/windowscodecs/libtiff.c
index 16701a258f9..ca5ccf3237a 100644
--- a/dlls/windowscodecs/libtiff.c
+++ b/dlls/windowscodecs/libtiff.c
@@ -780,7 +780,7 @@ static HRESULT tiff_decoder_read_tile(struct tiff_decoder *This, UINT tile_x, UI
 
         count = width_bytes * info->tile_height;
 
-        srcdata = malloc(count);
+        srcdata = malloc(count + 1);
         if (!srcdata) return E_OUTOFMEMORY;
         memcpy(srcdata, This->cached_tile, count);
 
@@ -819,7 +819,7 @@ static HRESULT tiff_decoder_read_tile(struct tiff_decoder *This, UINT tile_x, UI
 
         count = width_bytes * info->tile_height;
 
-        srcdata = malloc(count);
+        srcdata = malloc(count + 2);
         if (!srcdata) return E_OUTOFMEMORY;
         memcpy(srcdata, This->cached_tile, count);
 
-- 
2.47.1

